[quant] Fix fuse linear pass

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40549 [quant] Fix fuse linear pass**

Summary:
Currently we didn't check if %weight_t is produced by `aten::t`, this will fuse some `matmul`/`addmm` that is
not 2d to `aten::linear`, which is incorrect

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22225921](https://our.internmc.facebook.com/intern/diff/D22225921)

jerryzh168 committed:
[quant] Fix fuse linear pass

Summary:
Currently we didn't check if %weight_t is produced by `aten::t`, this will fuse some `matmul`/`addmm` that is
not 2d to `aten::linear`, which is incorrect

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Fix fuse linear pass"


Summary:
Currently we didn't check if %weight_t is produced by `aten::t`, this will fuse some `matmul`/`addmm` that is
not 2d to `aten::linear`, which is incorrect

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on torch/csrc/jit/passes/fuse_linear.cpp:
```diff
@@ -21,10 +21,22 @@ void FuseLinear(std::shared_ptr<Graph>& graph) {
     return is_int_constant(match, vmap, "beta", 1);
   };
 
+  // check %weight_t is produced by `aten::t` to make sure
```
This feels brittle, what if the user already has a weight that is transposed already?

jerryzh168 replied:
we don't transform it to `aten::linear` in that case since we don't know the dimension of weight. we can implement quantized::matmul if we need to quantize it

jerryzh168 committed:
Update on "[quant] Fix fuse linear pass"


Summary:
Currently we didn't check if %weight_t is produced by `aten::t`, this will fuse some `matmul`/`addmm` that is
not 2d to `aten::linear`, which is incorrect

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Fix fuse linear pass"


Summary:
Currently we didn't check if %weight_t is produced by `aten::t`, this will fuse some `matmul`/`addmm` that is
not 2d to `aten::linear`, which is incorrect

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Fix fuse linear pass"


Summary:
Currently we didn't check if %weight_t is produced by `aten::t`, this will fuse some `matmul`/`addmm` that is
not 2d to `aten::linear`, which is incorrect

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22225921](https://our.internmc.facebook.com/intern/diff/D22225921)

[ghstack-poisoned]
